### ISSUE TITLE
Latency Issues / Concurency Queue

### ISSUE THREAD

**Issue by KnutAtteslander** [cockroachdb/cockroach#25345]
Hi

Since a couple of days we have a sudden increase in query latency (~12..15s).
The table contains binary data and is around 88GB in size. However the size of this table ever increases and never drops. I changed the GC time to 3600s and deleted ~2/3 (15500/25k) of the rows (which where previously "softdeleted") ~12h later still no change in table size.
Querys on ID of this tables are normaly fast, but as soon as i include binary_data in the select query time increases to >10s.

I also have an increase in pending actions in the "Consistency Checker Queue"
[Imgur](https://imgur.com/a/yDtDh7G)

Related(?):
#21824
#15832

Cockroach runs on 4 nodes on 2 virtual machines on 2 Hyper-V hosts.
VM spec: 49152MB RAM, 127GB hd
VM2 (FVAPI-2) is around 80% disk usage, VM1 (FVAPI-1) is around 50%.

ID | Address | Uptime | Bytes | Replicas | Mem Usage | Version
n1 | db1.xxx.ch:26257 | 23 days | 23.3 GiB | 725 | 14.9 GiB | v2.0.0 
n3 | db3.xxx.ch:26257 | 23 days | 8.7 GiB | 683 | 14.1 GiB | v2.0.0 
n4 | db2.xxx.ch:26257 | 23 days | 28.5 GiB | 726 | 14.2 GiB | v2.0.0 
n5 | db4.xxx.ch:26257 | an hour | 28.5 GiB | 725 | 13.8 GiB | v2.0.0 

I180507 06:22:27.190192 1 util/log/clog.go:1104  [config] file created at: 2018/05/07 06:22:27
I180507 06:22:27.190192 1 util/log/clog.go:1104  [config] running on machine: fvapi-2
I180507 06:22:27.190192 1 util/log/clog.go:1104  [config] binary: CockroachDB CCL v2.0.0 (x86_64-unknown-linux-gnu, built 2018/04/03 20:56:09, go1.10)
I180507 06:22:27.190192 1 util/log/clog.go:1104  [config] arguments: [cockroach start --certs-dir=/home/user/certs --store=/home/user/node4 --host=db4.xxx.ch --port=26257 --http-port=8080 --locality=host=fvapi-2 --cache=25% --max-sql-memory=10% --join=db2.xxx.ch]

I180507 06:46:09.420969 297 util/log/clog.go:1001  [config] file created at: 2018/05/07 06:46:09
I180507 06:46:09.420969 297 util/log/clog.go:1001  [config] running on machine: fvapi-2
I180507 06:46:09.420969 297 util/log/clog.go:1001  [config] binary: CockroachDB CCL v2.0.0 (x86_64-unknown-linux-gnu, built 2018/04/03 20:56:09, go1.10)
I180507 06:46:09.420969 297 util/log/clog.go:1001  [config] arguments: [cockroach start --certs-dir=/home/user/certs --store=/home/user/node2 --host=db2.xxx.ch --port=26257 --http-port=8080 --locality=host=fvapi-2 --cache=25% --max-sql-memory=10% --join=db1.xxx.ch]

I180505 00:56:15.974823 168 util/log/clog.go:1001  [config] file created at: 2018/05/05 00:56:15
I180505 00:56:15.974823 168 util/log/clog.go:1001  [config] running on machine: fvapi-1
I180505 00:56:15.974823 168 util/log/clog.go:1001  [config] binary: CockroachDB CCL v2.0.0 (x86_64-unknown-linux-gnu, built 2018/04/03 20:56:09, go1.10)
I180505 00:56:15.974823 168 util/log/clog.go:1001  [config] arguments: [cockroach start --certs-dir=/home/user/certs --store=/home/user/node3 --host=db3.xxx.ch --port=26257 --http-port=8080 --locality=host=fvapi-1 --cache=25% --max-sql-memory=10% --join=db1.xxx.ch]

I180507 06:21:53.248366 177681051 util/log/clog.go:1001  [config] file created at: 2018/05/07 06:21:53
I180507 06:21:53.248366 177681051 util/log/clog.go:1001  [config] running on machine: fvapi-1
I180507 06:21:53.248366 177681051 util/log/clog.go:1001  [config] binary: CockroachDB CCL v2.0.0 (x86_64-unknown-linux-gnu, built 2018/04/03 20:56:09, go1.10)
I180507 06:21:53.248366 177681051 util/log/clog.go:1001  [config] arguments: [cockroach start --certs-dir=/home/user/certs --store=/home/user/node1 --host=db1.xxx.ch --port=26257 --http-port=8080 --locality=host=fvapi-1 --cache=25% --max-sql-memory=10%]

**Comment by KnutAtteslander** [comment]
Additional infos:
A sql dump of the whole database is ~4.2 GB before i massdeleted the rows
and ~1.3 GB after deleting the ~15k rows.

**Comment by tschottdorf** [comment]
Could you send me a SHOW TRACE FOR for suck a slow query? How big is binary_data?

**Comment by KnutAtteslander** [comment]
root@db1.fvapi.gblsrv.ch:26257/fvapi> SHOW TRACE FOR SELECT image_data FROM foto_informations WHERE id = '332210770908119041';
+----------------------------------+---------------+-------------------------------------------------------+--------------------------------------------------------+-----+-----------------------------------+------+
|            timestamp             |      age      |                        message                        |                          tag                           | loc |             operation             | span |
+----------------------------------+---------------+-------------------------------------------------------+--------------------------------------------------------+-----+-----------------------------------+------+
| 2018-05-07 10:23:29.471705+00:00 | 0s            | === SPAN START: sql txn ===                           |                                                        |     | sql txn                           |    0 |
| 2018-05-07 10:23:29.473232+00:00 | 1ms526µs906ns | === SPAN START: session recording ===                 |                                                        |     | session recording                 |    6 |
| 2018-05-07 10:23:29.473247+00:00 | 1ms541µs906ns | === SPAN START: starting plan ===                     |                                                        |     | starting plan                     |    1 |
| 2018-05-07 10:23:29.473284+00:00 | 1ms578µs906ns | === SPAN START: consuming rows ===                    |                                                        |     | consuming rows                    |    2 |
| 2018-05-07 10:23:29.473343+00:00 | 1ms638µs507ns | Scan /Table/53/1/332210770908119041{-/#}              | [n4,client=85.195.226.26:49617,user=root]              |     | sql txn                           |    0 |
| 2018-05-07 10:23:29.473377+00:00 | 1ms672µs707ns | === SPAN START: dist sender ===                       |                                                        |     | dist sender                       |    3 |
| 2018-05-07 10:23:29.47341+00:00  | 1ms705µs7ns   | querying next range at /Table/53/1/332210770908119041 | [client=85.195.226.26:49617,user=root,txn=dd6687d2,n4] |     | dist sender                       |    3 |
| 2018-05-07 10:23:29.473507+00:00 | 1ms802µs607ns | r145: sending batch 1 Scan to (n3,s3):7               | [client=85.195.226.26:49617,user=root,txn=dd6687d2,n4] |     | dist sender                       |    3 |
| 2018-05-07 10:23:29.473588+00:00 | 1ms883µs508ns | sending request to db3.fvapi.gblsrv.ch:26257          | [client=85.195.226.26:49617,user=root,txn=dd6687d2,n4] |     | dist sender                       |    3 |
| 2018-05-07 10:23:29.473634+00:00 | 1ms929µs408ns | === SPAN START: /cockroach.roachpb.Internal/Batch === |                                                        |     | /cockroach.roachpb.Internal/Batch |    4 |
| 2018-05-07 10:23:29.474752+00:00 | 3ms47µs497ns  | === SPAN START: /cockroach.roachpb.Internal/Batch === |                                                        |     | /cockroach.roachpb.Internal/Batch |    5 |
| 2018-05-07 10:23:29.474822+00:00 | 3ms116µs898ns | 1 Scan                                                | [n3]                                                   |     | /cockroach.roachpb.Internal/Batch |    5 |
| 2018-05-07 10:23:29.474847+00:00 | 3ms142µs598ns | executing 1 requests                                  | [n3,s3]                                                |     | /cockroach.roachpb.Internal/Batch |    5 |
| 2018-05-07 10:23:29.474868+00:00 | 3ms163µs98ns  | read-only path                                        | [n3,s3,r145/7:/Table/53/1/33{17091.-22806.}]           |     | /cockroach.roachpb.Internal/Batch |    5 |
| 2018-05-07 10:23:29.474887+00:00 | 3ms182µs698ns | command queue                                         | [n3,s3,r145/7:/Table/53/1/33{17091.-22806.}]           |     | /cockroach.roachpb.Internal/Batch |    5 |
| 2018-05-07 10:23:29.474906+00:00 | 3ms200µs798ns | waiting for read lock                                 | [n3,s3,r145/7:/Table/53/1/33{17091.-22806.}]           |     | /cockroach.roachpb.Internal/Batch |    5 |
| 2018-05-07 10:23:29.475201+00:00 | 3ms496µs400ns | read completed                                        | [n3,s3,r145/7:/Table/53/1/33{17091.-22806.}]           |     | /cockroach.roachpb.Internal/Batch |    5 |
| 2018-05-07 10:23:29.478402+00:00 | 6ms697µs126ns | plan completed execution                              | [n4,client=85.195.226.26:49617,user=root]              |     | consuming rows                    |    2 |
| 2018-05-07 10:23:29.478406+00:00 | 6ms701µs626ns | resources released, stopping trace                    | [n4,client=85.195.226.26:49617,user=root]              |     | consuming rows                    |    2 |
+----------------------------------+---------------+-------------------------------------------------------+--------------------------------------------------------+-----+-----------------------------------+------+
(19 rows)

Time: 292.2671ms

binary data (image_data) is between 500kb and 800kb

**Comment by tschottdorf** [comment]
Do you have one that shows the 10+s? 

@jordanlewis where do the 200ms go in the above trace, is that expected?

**Comment by KnutAtteslander** [comment]
Unfortuneatly not :/
The querys where not performed on the cluster itself!

If i do a 
`cockroach sql --execute="SELECT image_data FROM foto_informations WHERE id = '332210770908119041';" --host=db1.xxx.ch --certs-dir=certs > /mnt/backup/dump.txt`
the querry is done in ~1...2s

Below an image from the metrics view:
[Imgur](https://imgur.com/a/ANoZrFi)

Another image with the time marked when the problems startet:
[Imgur2](https://imgur.com/a/9GPlV2u)

Further inspection of the cluster, revealed 2 dead nodes; those are from a previous attempt to join two additional nodes to the cluster in a different location, which didnt work because of timing issues (probably poor internet connection). These nodes can't be decommissioned (Error: rpc error: code = Unknown desc = unable to get liveness for 9: node not in the liveness table) and therefore no node info (cockroach node status) available.

I will now: 
1) create a new cluster in dev environment and populate with a backup.
2) check the querys made by our application on dev
3) create new production cluster and populate with backup

**Comment by tschottdorf** [comment]
@KnutAtteslander thanks for following up. To clarify, did you ever see queries actually taking 10+ seconds or are you only worried about the metric? There's definitely a problem here. If you still have that cluster running, could you create a `./cockroach debug zip` on each node and send them to me? You can use my email, which is tobias with domain cockroachlabs.com. (The debug logs can leak individual primary keys in the logs, so they're sensitive).

**Comment by tschottdorf** [comment]
(oh and if you can trace a slow query -- that will be very useful).

**Comment by tschottdorf** [comment]
@KnutAtteslander, do you have an update here?

**Comment by KnutAtteslander** [comment]
Sorry for the delay.
I was able to initialize a new cluster, which works fine since, and restore the backup.
The metrics on the consistancy queue seems fine -> no pending action so far.
The latency is gone also.
**My immediate problems are solved so far.**

I have some issues with the database size -> the live bytes are around 25 GB (capacity at 60 GB) contrasting the database backup, which is 3.5 GB.
I see a dayily dip in the live bytes, due to the delete actions, but the trend is climbing.
As my problems arised the database was around 90 GB and on one node consuming >80% of the filesystem. I resized the filesystem to mitigate future problems there.
The cockroach zone configs are now on default settings, runnig version 2.0.1

I can try to setup a new cluster with the old datastores and make a debug dump if you like.
Howerver, if i do i have to exclude the big table (99% of database size), if possible, because the data in there is very sensitive personal information.

**Comment by KnutAtteslander** [comment]
I hate to say this 2 days after i said "my immediate problems are solved so far" but the pending actions in the consistency queue are rising again...
I will reinitialize the cluster, again, next weekend. I'd rather not do this, but right now it's the only action that "solves" the latency issues that will emerge again.

What i don't understand is; i can make time travel queries on one table as far back as 2018-05-24 on one table and on the huge one as far back as 2018-05-27 -> but the gc-ttlseconds is set to 90'000s (default settings).

I just done some time travel querys and found some pretty old data (soft-deleted 2018-05-16), which may contribute to the huge database size. **Is it possible to completly disable time travel functionality or reduce it to an absolute minimum?**

**Comment by KnutAtteslander** [comment]
Yes by soft-deleted i mean we have a deleted_at field (gorm.io golang package).
We run a script, once a day, which checks the deleted_at column and rows "older" than 10 days will be deleted with a DELETE statement.
This works, as if do a non time travel query i don't get results older than 10 days.

Removing 2/3rds of the rows didn't changed a thing last time, apart from reducing the backup sice.

With live bytes i refer to the storage dashboard metrics.
The backup is made with a sql dump, we don't habe an enterprise licence (Alltough that is something i want to discuss with my boss. In the end we make money using your product).

Ok so i don't have t worry about the pending actions.

I will do that experiment in the next days and report back here. But first i need to do some preparations for the reinitialisation of the cluster next week.

Thanks for your support!

**Comment by tschottdorf** [comment]
> But first i need to do some preparations for the reinitialisation of the cluster next week.

Don't let that stop you from preparing this as a fallback option, but the intention is definitely that you do not have to set that cluster up anew.

Thanks for letting me take a look at your cluster's dashboard (private communication).

As you mentioned I'm seeing live bytes at around 30gb. This over-counts your "real data" by a factor of 3 since you run with a replication factor of 3 (i.e. each byte is duplicated to two further nodes). So you really have ~10gb of unique live bytes and I would have expected this to roughly correspond to the size of a SQL dump. You mention that a SQL dump is around 3-4GB. Are you compressing that dump by any chance? You say that you have 25k rows, if each has an 500kb TIFF image (and that accounts for basically all of the size) then you end up with 12.5GB of raw TIFF image data, which is in the ballpark of my computations above, but I don't see how that can become a 3-4GB backup unless you're compressing it well (or I have the wrong number of rows?).

For the 10GB of logical bytes, taking replication into account again, each node holds on to around 30/4 = 7.5GB of logical live data on their disks.
The actual approximate disk usage for that table at 60gb, as you reported. This amounts to 15-20gb of disk per node. I think this roughly checks out, because:

1. each node has around 7.5gb of live bytes
1. RocksDB has write amplification, i.e. you end up with an average of, say, 3x your logical size on disk unless the data is freshly compacted and doesn't change much.
1. RocksDB compresses data on-disk, which saves you some factor. Since the bulk of your payload is TIFF, it depends on whether you store it in compressed form. The factor could be anywhere between 1.0 and 0.1. Your nodes have 3x20 and 1x7 gb, respectively, of disk usage.

**Comment by tschottdorf** [comment]
Going out on a limb here, let's assume the following:

1. the backup is really ~4gb
1. the table really has ~25k rows and that hasn't changed recently
1. the live bytes is really ~30gb and corresponds to the on-disk storage convincingly

Then the remaining question is: why are there so many live bytes?

**Comment by gpaul** [comment]
If there are some (several?) indexes this might account for the difference.

**Comment by KnutAtteslander** [comment]
- Image_data is JPG.
- There are several indexes.
- The SQL dump is uncompressed txt.

I have to correct the following statement of mine: 
- Image_data is around ~100KB.
I was told image_data is around 800..1000KB, but I made a dump of 100 rows wich resulted in an uncompressed txt file around 10MB.

**Comment by tschottdorf** [comment]
We looked at the table a little more and there are 28000 records, with image size ~100kb resulting in an expected payload of ~3gb, which is in line with the size of the dump.

@gpaul that's a good point, the table has three secondary indexes, but none of them contain the blob so they *should* be "negligible" and in particular, when you delete a row, they also get deleted.

Also, looking at the UI, I see a graph like this:

<img width="970" alt="image" src="https://user-images.githubusercontent.com/5076964/40787322-51df2f22-64bb-11e8-980c-519b0113a321.png">

During this time, the table contains about the same number of entries -- 25-28k. So it definitely seems like something is adding LiveBytes over time.

**Comment by tschottdorf** [comment]
Looking through `crdb_internal.ranges` and the range status page, I can check out individual ranges' MVCCStats. One curiosity is that of the ~500 ranges of the big table, approximately the first half is empty. Are you by any chance using monotonic primary keys (such as a timestamp)?

Another issue I see (though it doesn't at all explain the behavior here) is this:

<img width="456" alt="image" src="https://user-images.githubusercontent.com/5076964/40799194-eb521786-64da-11e8-92bd-a0c1ab31af43.png">

There are 66 live keys. Since you have everything in one column family, this means this particular range holds 66 SQL rows. The middle row in the image shows how many rows and (SQL-)deleted rows you have: also 66. But the last row shows a count of 323 values, which basically means that these 66 rows have been updated a bunch of times (323-66 times in total). Each time you update such a row, even if you're just changing a small part of it, you're rewriting the image.

To deal with this, you're going to want to put the image into its own [column family](https://www.cockroachlabs.com/docs/stable/column-families.html)

But this is just a tangent -- it doesn't explain the live bytes we see here. Older versions of rows don't show up in LiveBytes.

In this graph below, we see live bytes, key bytes and val bytes together. Key bytes is barely visible -- the keys are very small, so they don't really count for much since the values get so large. The interesting takeaway here is that periodically, ValBytes dips down so that it basically equals LiveBytes. I think these are your nightly deletions -- they actually manage to delete almost all of the old data (il.e. deleted rows). But there is something else -- something that contributes LiveBytes but isn't in that table. If we found out what that is, and subtracted it, the remaining graph for your table would be bounded by ~3gb.

<img width="1433" alt="image" src="https://user-images.githubusercontent.com/5076964/40800872-890e3fa0-64df-11e8-9b3f-634b0dd7c1d9.png">

In conclusion, we remain looking for the contributor of LiveBytes. One possible source are time series data -- we don't compute those correctly and I believe they are counted in live bytes.

@mrtracy, can you comment here? Timeseries are live bytes, right (?). We overcount time series data's bytes, but when we delete, we probably subtract the real value, leaving an eternal delta that would only be corrected when the consistency checker runs (but that is blocked on this cluster). Can this amount to the difference?

**Comment by tschottdorf** [comment]
@KnutAtteslander I looked through this with my colleague @mrtracy and I think we've arrived at an explanation for the ever-growing live bytes: it might be time series after all! We computed that you'd expect about 1GB per day of logical growth in your cluster. Looking at the live bytes, that's extremely exactly what you have. Next, we checked out the logical size of your large table, and it came out to 18.2gb -- your cluster started around May 8, so that checks out (roughly). The ui shows only ~3gb of on-disk storage. It's likely that the timeseries data compresses extremely well (since much of it is zero).

There's still a bit of a discrepancy: you have 30gb of live bytes, 18 of that in timeseries. That leaves 12gb to explain. We looked at your large table; it's 6gb of live bytes. That leaves 6gb still that we'll have to think about.

But, for today, I'd like you to `SET CLUSTER SETTING timeseries.resolution_10s.storage_duration = '168h'` which means that timeseries data older than a week will be deleted (so anything older than that will just disappear). This should give you a noticeable drop in live bytes and hopefully we'll observe that the growth you've seen stops (it would've stopped naturally after 30 days, but I don't think you want to wait for another week to find out if that's really true).

Feel free to adjust that number if you need more than a week of history, but it should be below 18.

**Comment by tschottdorf** [comment]
And my other suggestion about running a manual compaction still stands. I think that your delete-heavy workload isn't sufficiently optimized for on our end, which results in lots of disk space not being released in a timely manner. That will hopefully alleviate the disk usage which I think is mostly driven by your large table (not time series).

**Comment by KnutAtteslander** [comment]
I have set the timeseries settings according to your recommondation.
Tomorow i will go through your other suggestions. 
Thank you so far!

**Comment by KnutAtteslander** [comment]
The Live Bytes are gradualy dropping since 9h ago.
From 30.3GB to 25.4GB (now). I maybe further reduce the timeseries period, I really only need live data for 1 maybe 2 days.
For longer periods the prometheus/grafana data should be enough (collecting metrics every 5 minutes), but at the moment prometheus doesn't collect all metrics i have to adjust that first.

I will also adjust the column families as you mentioned, rows in the foto_informations table don't get that much updates but there's no reason not to prevent rewriting the blob.
But for that i have to reinitialize the cluster; as far as i understand it, i can not rearange existing columns.

The manual compaction i will do on this sunday, so that i dont impact production if i break something 😄 
If that turns out to reduce disk size, we will incorporate that in a weekly maintenance action.

**Comment by tschottdorf** [comment]
Thanks @KnutAtteslander! Yes -- you need to reinitialize when you're changing the column families. You may chose to leave it as-is for now, the problem doesn't seem to be too bad in your cluster, but having rows which see lots of updates -- even if it's just one row that gets updated over and over by accident - may cause you to use a lot more disk space than you had anticipated

By the way, the plan for CockroachDB v2.1 is that you wouldn't have to run this manual compaction regularly; this should really not require manual intervention.

**Comment by KnutAtteslander** [comment]
Production just informed me that they experienced a significant decrease in latency since this morning 👍 
Live Bytes are below 25GB, it almost seems like thats some sort of a threashold.

**Comment by tschottdorf** [comment]
@KnutAtteslander I'm going to close this issue. I'm happy to take another look at the cluster in a bit to check on the disk space situation. You should also run the compaction (you can report here even if the issue is closed).
As new problems come up, please just open a new issue. And thanks for providing all that information!

**Comment by KnutAtteslander** [comment]
Hi

I did the compaction:

$ du -hs /home/node2
26G     /home/node2

$ cockroach debug compact /home/node2
approximate reported database size before compaction: 26 GiB
approximate reported database size after compaction: 324 MiB

$ du -hs /home/node2
436M    /home/node2

----------------------------------------

$ du -hs /home/node4
6.9G    /home/node4

$ cockroach debug compact /home/node4
approximate reported database size before compaction: 6.7 GiB
approximate reported database size after compaction: 468 MiB

$ du -hs /home/node4
568M    /home/node4

----------------------------------------

$ du -hs /home/node1
27G     /home/node1

$ cockroach debug compact /home/node1
approximate reported database size before compaction: 26 GiB
approximate reported database size after compaction: 353 MiB

$ du -hs /home/node1
455M    /home/node1

----------------------------------------

$ du -hs /home/node3
25G     /home/node3

$ cockroach debug compact /home/node3
approximate reported database size before compaction: 24 GiB
approximate reported database size after compaction: 335 MiB

$ du -hs /home/node3
452M    /home/node3

**Comment by KnutAtteslander** [comment]
I deleted all the "soft-deleted" rows in the database before compacting -> The database is now 190**M**B.

The RocksDB SSTables went down from 1866 to 41

![after_compaction](https://user-images.githubusercontent.com/894341/41199528-00359b08-6c94-11e8-8449-d732082a13b1.png)

**Comment by tschottdorf** [comment]
Nice, that went well. We'll have to look into making that less of a manual requirement.

**Comment by KnutAtteslander** [comment]
Yeah i'm looking forward to the next version.
Thanks again for the awesome help!
